Options-struct interfaces for creating publishers/subscribers (pre-QoS, standalone)

emersonknapp opened the pull request:
Adds new PublisherOptions and SubscriptionOptions classes, with new Publisher and Subscriber constructors to accept them.

QoS event callbacks will be put into these Options structs - so this is the only API change needed for QoS event callbacks.

Signed-off-by: Emerson Knapp <eknapp@amazon.com>

gbiggs commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -139,7 +139,7 @@ class Node : public std::enable_shared_from_this<Node>
   get_callback_groups() const;
 
-  /// Create and return a Publisher.
+  /// Create and return a Publisher. Note: This constructor is deprecated
```
Does the [deprecate attribute](https://en.cppreference.com/w/cpp/language/attributes/deprecated) work for constructors?

emersonknapp replied:
The attribute should work (my comment may be misnaming, this are factory functions, not contsructors) - My one concern with marking these as deprecated is that it will throw warnings in other parts of the core until they are migrated, which will cause the CI builds to be marked as unstable - I'm not sure if we want to do that yet.

wjwwood replied:
> My one concern with marking these as deprecated is that it will throw warnings in other parts of the core until they are migrated, which will cause the CI builds to be marked as unstable - I'm not sure if we want to do that yet.

Commenting that they are deprecated without deprecating them using the attribute is definitely not what we want to do, it should either be removed, or deprecated with the attribute (with or without a comment in the doc block), or not deprecated at all.

I think we should deprecate it with the attribute, and look at the downstream impact (how many places need to be updated). If the number is low, then we can consider just removing it directly.

@emersonknapp can you please collect a list of all the places that need to be touched in code listed in the `ros2.repos` file? You could do this by deprecating this interface and running CI, which will collect the compiler warnings for you in one place.

As we also spoke about off-line, I do think we should address the issue of removing or deprecating the existing interfaces and fixing any new downstream failures or warnings due to that all at once, rather than a two stage process where we merge the new API and then separately get rid of the other ones.

wjwwood replied:
This applies to all the interfaces here that may need to be deprecated.

emersonknapp replied:
Good plan, I will go ahead and make those changes and run CI

gbiggs commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -151,10 +151,11 @@ class Node : public std::enable_shared_from_this<Node>
     std::shared_ptr<Alloc> allocator = nullptr);
 
-  /// Create and return a Publisher.
+  /// Create and return a Publisher. Note: this constructor is deprecated
```
Use deprecate attribute if possible.

gbiggs commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -170,7 +171,23 @@ class Node : public std::enable_shared_from_this<Node>
+    const PublisherOptions<Alloc> & options,
+    rclcpp::callback_group::CallbackGroup::SharedPtr callback_group = nullptr);
+
+  /// Create and return a Subscription. Note: this constructor is deprecated
```
Use deprecate attribute if possible.

gbiggs commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -203,7 +220,7 @@ class Node : public std::enable_shared_from_this<Node>
     std::shared_ptr<Alloc> allocator = nullptr);
 
-  /// Create and return a Subscription.
+  /// Create and return a Subscription. Note: this constructor is deprecated
```
Use deprecate attribute if possible.

wjwwood commented on rclcpp/include/rclcpp/callback_group.hpp:
```diff
@@ -22,6 +22,7 @@
 
 #include "rclcpp/client.hpp"
 #include "rclcpp/service.hpp"
+#include "rclcpp/publisher.hpp"
```
alphabetical ordering

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -170,7 +171,23 @@ class Node : public std::enable_shared_from_this<Node>
+  std::shared_ptr<PublisherT>
+  create_publisher(
+    const std::string & topic_name,
+    const PublisherOptions<Alloc> & options,
```
This should have a default argument, right?

wjwwood replied:
Also, I think we ought to keep the `size_t qos_history_depth` as a separate, non-default argument in this method.

This is related to another issue, about whether or not we should always force users to specify the history depth. But having a another signature that doesn't require you to specify it is a step backwards I think.

emersonknapp replied:
> This should have a default argument, right?

It couldn't be defaulted while keeping the other signatures, since it's `create_publisher(topic_name)` becomes ambiguous in that case. 

I'll go ahead with the analysis on which other packages need to be changed, which informs how this looks anyways.

> we ought to keep the size_t qos_history_depth as a separate, non-default argument in this method

Ok - I wasn't aware before that we wanted to force users to always specify. I will change in that direction

emersonknapp replied:
@wjwwood one question for forcing qos_history_depth: which should have precedence, the function argument or the value in the qos_profile in the Options struct?

wjwwood replied:
The function argument.

Later, I may have a QoS factory function that requires you to specify the depth to get a QoS, and then make the QoS a required argument, but for now, requiring the depth and overriding the value in the default QoS structure is the way to go.

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -600,6 +645,16 @@ class Node : public std::enable_shared_from_this<Node>
   const NodeOptions &
   get_node_options() const;
 
+  /// Manually assert that this Node is alive (for RMW_QOS_POLICY_MANUAL_BY_NODE)
```
nitpick: punctuation

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -236,6 +253,34 @@ class Node : public std::enable_shared_from_this<Node>
+  create_subscription(
+    const std::string & topic_name,
+    CallbackT && callback,
+    const SubscriptionOptions<Alloc> & options,
```
Same comment here about not forcing users to specify an options struct, but forcing them to specify a history depth.

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -600,6 +645,16 @@ class Node : public std::enable_shared_from_this<Node>
+  /**
+   * If the rmw Liveliness policy is set to RMW_QOS_POLICY_MANUAL_BY_NODE, the creator of this
+   * Node must manually call `assert_liveliness` periodically to signal that this Node
+   * is still alive. Must be called at least as often as qos_profile's Liveliness lease_duration
```
nitpick: one sentence per line

wjwwood replied:
also nitpick: punctuation and complete sentences

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -600,6 +645,16 @@ class Node : public std::enable_shared_from_this<Node>
+   */
+  RCLCPP_PUBLIC
+  void
+  assert_liveliness() {}
```
This should be implemented in `node.cpp` and not in this header, and also implemented at some point :)

wjwwood commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -31,17 +31,21 @@
-// Forward declaration is used for friend statement.
 namespace node_interfaces
 {
+// NOTE(emersonknapp) Forward declaration avoids including node_base_interface.hpp which causes
```
No need for `NOTE` (it is flagged by some linters as something to address, similar to a TODO and is unnecessary).

wjwwood commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -128,6 +132,16 @@ class PublisherBase
   size_t
   get_intra_process_subscription_count() const;
 
+  /// Manually assert that this Publisher is alive (for RMW_QOS_POLICY_MANUAL_BY_TOPIC)
```
nitpick: punctuation

wjwwood replied:
I see this a few times elsewhere too, but I won't comment on each, please skim over the docblocks at some point and make sure complete sentences have punctuation.

wjwwood commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -128,6 +132,16 @@ class PublisherBase
+  /**
+   * If the rmw Liveliness policy is set to RMW_QOS_POLICY_MANUAL_BY_TOPIC, the creator of this
+   * Publisher must manually call `assert_liveliness` periodically to signal that this Publisher
+   * is still alive. Must be called at least as often as qos_profile's Liveliness lease_duration
```
nitpick: one sentence per line

wjwwood commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -128,6 +132,16 @@ class PublisherBase
+   */
+  RCLCPP_PUBLIC
+  void
+  assert_liveliness() {}
```
Since this is in the base class (non-templated), it should go into `publisher.cpp`.

wjwwood commented on rclcpp/include/rclcpp/publisher_options.hpp:
```diff
@@ -0,0 +1,44 @@
+{
+
+/// Contains callbacks for various types of events a Publisher can receive from the middleware
+struct PublisherEventCallbacks
```
This should have its own file (if it is left in this pull request, I'll comment on the scope of changes here at the end).

wjwwood commented on rclcpp/include/rclcpp/subscription_options.hpp:
```diff
@@ -0,0 +1,44 @@
+{
+
+/// Contains callbacks for non-message events that a Subscriber can receive from the middleware
+struct SubscriptionEventCallbacks
```
Should have its own file.

emersonknapp commented:
@thomas-moulard - please run the following CI job:

 * Gist file: https://gist.githubusercontent.com/emersonknapp/75df97702d35ef941742db8bd0f84174/raw/2391ac7887abd36240fe8a9e0e564f2b43f0b65c/ros2.repos
 * Additional test flags: --packages-up-to rclcpp
 * CI Job: ci_launcher

emersonknapp commented:
@wjwwood CI pending to see the impact downstream - took recommendations and laser-focused to the Options interfaces
* removed callbackgroups for publishers
* removed EventCallbacks placeholders

I noticed that the `create_*(topic)`  version defaults to the `create_*(topic, qos_profile)` version of these interfaces, since `qos_history_depth` was not mandatory there - so just within the tests for this package we'll see a lot of warnings from a `[[deprecated]]` version. I left those tests out of this PR just so we can count up the total impact of forcing `qos_history_depth` on the API, from the CI run

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6629)](http://ci.ros2.org/job/ci_linux/6629/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2955)](http://ci.ros2.org/job/ci_linux-aarch64/2955/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5439)](http://ci.ros2.org/job/ci_osx/5439/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6451)](http://ci.ros2.org/job/ci_windows/6451/)

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6630)](http://ci.ros2.org/job/ci_linux/6630/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2956)](http://ci.ros2.org/job/ci_linux-aarch64/2956/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5440)](http://ci.ros2.org/job/ci_osx/5440/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6452)](http://ci.ros2.org/job/ci_windows/6452/)

thomas-moulard commented:
Cancelled the other job and started new ones building and running everything.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6632)](http://ci.ros2.org/job/ci_linux/6632/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2958)](http://ci.ros2.org/job/ci_linux-aarch64/2958/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5442)](http://ci.ros2.org/job/ci_osx/5442/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6454)](http://ci.ros2.org/job/ci_windows/6454/)

emersonknapp commented:
@thomas-moulard it looks like the build failed upstream because I was behind `ros2/master` - my branch didn't have `rclcpp_components` yet

dirk-thomas commented:
> it looks like the build failed ...

If the builds are failing anyway please consider aborting pending builds to not waste cycles on the Jenkins nodes.

thomas-moulard commented:
@dirk-thomas Will do next time, I was in meetings when it happened.

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6637)](http://ci.ros2.org/job/ci_linux/6637/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2963)](http://ci.ros2.org/job/ci_linux-aarch64/2963/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5447)](http://ci.ros2.org/job/ci_osx/5447/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6460)](http://ci.ros2.org/job/ci_windows/6460/)

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6643)](http://ci.ros2.org/job/ci_linux/6643/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2969)](http://ci.ros2.org/job/ci_linux-aarch64/2969/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5455)](http://ci.ros2.org/job/ci_osx/5455/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6468)](http://ci.ros2.org/job/ci_windows/6468/)

wjwwood commented on rclcpp/include/rclcpp/node_impl.hpp:
```diff
@@ -113,6 +101,35 @@ Node::create_publisher(
+    topic_name, qos_history_depth, pub_options);
+}
+
+
```
nitpick: minimize vertical whitespace (https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace)

emersonknapp replied:
fixed

wjwwood commented:
Looks like you have an ambiguous call:

```
10:59:22 /home/jenkins-agent/workspace/ci_linux/ws/src/ros2/system_tests/test_rclcpp/test/test_subscription.cpp: In member function ‘virtual void test_subscription__rmw_fastrtps_cpp_create_subscription_with_queue_size_Test::TestBody()’:
10:59:22 /home/jenkins-agent/workspace/ci_linux/ws/src/ros2/system_tests/test_rclcpp/test/test_subscription.cpp:383:38: error: call of overloaded ‘create_subscription(const char [18], test_subscription__rmw_fastrtps_cpp_create_subscription_with_queue_size_Test::TestBody()::<lambda(test_rclcpp::msg::UInt32_<std::allocator<void> >::ConstSharedPtr)>&, int)’ is ambiguous
10:59:22      "test_subscription", callback, 10);
10:59:22                                       ^
10:59:22 In file included from /home/jenkins-agent/workspace/ci_linux/ws/install/rclcpp/include/rclcpp/node.hpp:699:0,
10:59:22                  from /home/jenkins-agent/workspace/ci_linux/ws/install/rclcpp/include/rclcpp/executors/single_threaded_executor.hpp:28,
10:59:22                  from /home/jenkins-agent/workspace/ci_linux/ws/install/rclcpp/include/rclcpp/executors.hpp:22,
10:59:22                  from /home/jenkins-agent/workspace/ci_linux/ws/install/rclcpp/include/rclcpp/rclcpp.hpp:144,
10:59:22                  from /home/jenkins-agent/workspace/ci_linux/ws/src/ros2/system_tests/test_rclcpp/test/test_subscription.cpp:23:
10:59:22 /home/jenkins-agent/workspace/ci_linux/ws/install/rclcpp/include/rclcpp/node_impl.hpp:139:1: note: candidate: std::shared_ptr<SubscriptionT> rclcpp::Node::create_subscription(const string&, CallbackT&&, size_t, const rclcpp::SubscriptionOptions<Alloc>&, typename rclcpp::message_memory_strategy::MessageMemoryStrategy<typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>::SharedPtr) [with MessageT = test_rclcpp::msg::UInt32_<std::allocator<void> >; CallbackT = test_subscription__rmw_fastrtps_cpp_create_subscription_with_queue_size_Test::TestBody()::<lambda(test_rclcpp::msg::UInt32_<std::allocator<void> >::ConstSharedPtr)>&; Alloc = std::allocator<void>; SubscriptionT = rclcpp::Subscription<test_rclcpp::msg::UInt32_<std::allocator<void> >, std::allocator<void> >; std::__cxx11::string = std::__cxx11::basic_string<char>; size_t = long unsigned int; typename rclcpp::message_memory_strategy::MessageMemoryStrategy<typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>::SharedPtr = std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<test_rclcpp::msg::UInt32_<std::allocator<void> >, std::allocator<void> > >]
10:59:22  Node::create_subscription(
10:59:22  ^~~~
10:59:22 /home/jenkins-agent/workspace/ci_linux/ws/install/rclcpp/include/rclcpp/node_impl.hpp:227:1: note: candidate: std::shared_ptr<SubscriptionT> rclcpp::Node::create_subscription(const string&, CallbackT&&, size_t, rclcpp::callback_group::CallbackGroup::SharedPtr, bool, typename rclcpp::message_memory_strategy::MessageMemoryStrategy<typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>::SharedPtr, std::shared_ptr<PublisherT>, rclcpp::IntraProcessSetting) [with MessageT = test_rclcpp::msg::UInt32_<std::allocator<void> >; CallbackT = test_subscription__rmw_fastrtps_cpp_create_subscription_with_queue_size_Test::TestBody()::<lambda(test_rclcpp::msg::UInt32_<std::allocator<void> >::ConstSharedPtr)>&; Alloc = std::allocator<void>; SubscriptionT = rclcpp::Subscription<test_rclcpp::msg::UInt32_<std::allocator<void> >, std::allocator<void> >; std::__cxx11::string = std::__cxx11::basic_string<char>; size_t = long unsigned int; rclcpp::callback_group::CallbackGroup::SharedPtr = std::shared_ptr<rclcpp::callback_group::CallbackGroup>; typename rclcpp::message_memory_strategy::MessageMemoryStrategy<typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>::SharedPtr = std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<test_rclcpp::msg::UInt32_<std::allocator<void> >, std::allocator<void> > >]
10:59:22  Node::create_subscription(
10:59:22  ^~~~
```

CI is really slow, I'd recommend building locally before trying to see CI pass (it should really be for catching corner cases with your system or issues with OS's you don't have, not for trial and error on Linux).

emersonknapp commented:
@wjwwood I have pushed another commit removing optional parameters where they made the signature ambiguous. From my local run building the 233 packages - here are the places where the `-Wdeprecated-declarations` warning was triggered. Note - `rviz_default_plugins` enables `-Werror` and therefore would not build - I removed that single flag from it for the purpose of gathering these stats.

Result:
259 warnings in
117 files in
22 packages in
9 repos


```
ros2/rclcpp/rclcpp/test/test_time_source.cpp:133
ros2/rclcpp/rclcpp/test/test_publisher.cpp:75
ros2/rclcpp/rclcpp/test/test_publisher.cpp:80
ros2/rclcpp/rclcpp/test/test_publisher.cpp:100
ros2/rclcpp/rclcpp/test/test_publisher.cpp:111
ros2/rclcpp/rclcpp/test/test_publisher.cpp:117
ros2/rclcpp/rclcpp/test/test_publisher.cpp:124
ros2/rclcpp/rclcpp/test/test_pub_sub_option_interface.cpp:49
ros2/rclcpp/rclcpp/test/test_pub_sub_option_interface.cpp:60
ros2/rclcpp/rclcpp/test/test_publisher_subscription_count_api.cpp:77
ros2/rclcpp/rclcpp/test/test_publisher_subscription_count_api.cpp:82
ros2/rclcpp/rclcpp/test/test_publisher_subscription_count_api.cpp:94
ros2/rclcpp/rclcpp/test/test_subscription.cpp:98
ros2/rclcpp/rclcpp/test/test_subscription.cpp:115
ros2/rclcpp/rclcpp/test/test_subscription.cpp:128
ros2/rclcpp/rclcpp/test/test_subscription.cpp:133
ros2/rclcpp/rclcpp/test/test_subscription.cpp:147
ros2/rclcpp/rclcpp/test/test_subscription.cpp:152
ros2/rclcpp/rclcpp/test/test_subscription.cpp:157
ros2/rclcpp/rclcpp/test/test_subscription.cpp:165
ros2/rclcpp/rclcpp/test/test_subscription.cpp:190
ros2/rclcpp/rclcpp/test/test_subscription_publisher_count_api.cpp:68
ros2/rclcpp/rclcpp/test/test_subscription_publisher_count_api.cpp:72
ros2/rclcpp/rclcpp/test/test_subscription_publisher_count_api.cpp:81
ros2/rclcpp/rclcpp/test/test_serialized_message_allocator.cpp:60
ros2/demos/dummy_robot/dummy_map_server/src/dummy_map_server.cpp:36
ros2/demos/dummy_robot/dummy_sensors/src/dummy_laser.cpp:42
ros2/demos/dummy_robot/dummy_sensors/src/dummy_joint_states.cpp:33
ros2/rclcpp/rclcpp_action/test/test_client.cpp:199
ros2/rclcpp/rclcpp_action/test/test_client.cpp:208
ros2/rclcpp/rclcpp_action/test/test_server.cpp:246
ros2/rclcpp/rclcpp_action/test/test_server.cpp:307
ros2/rclcpp/rclcpp_action/test/test_server.cpp:362
ros2/rclcpp/rclcpp_action/test/test_server.cpp:419
ros2/rclcpp/rclcpp_action/test/test_server.cpp:474
ros2/rclcpp/rclcpp_action/test/test_server.cpp:530
ros2/demos/demo_nodes_cpp_native/src/talker.cpp:51
ros2/realtime_support/tlsf_cpp/example/allocator_example.cpp:77
ros2/realtime_support/tlsf_cpp/example/allocator_example.cpp:82
ros2/realtime_support/tlsf_cpp/test/test_tlsf.cpp:299
ros2/realtime_support/tlsf_cpp/test/test_tlsf.cpp:357
ros2/realtime_support/tlsf_cpp/test/test_tlsf.cpp:392
ros2/realtime_support/tlsf_cpp/test/test_tlsf.cpp:299
ros2/realtime_support/tlsf_cpp/test/test_tlsf.cpp:357
ros2/realtime_support/tlsf_cpp/test/test_tlsf.cpp:392
ros2/examples/rclcpp/minimal_publisher/lambda.cpp:31
ros2/examples/rclcpp/minimal_publisher/not_composable.cpp:31
ros2/examples/rclcpp/minimal_publisher/member_function.cpp:30
ros2/demos/logging_demo/src/logger_usage_component.cpp:34
ros2/examples/rclcpp/minimal_subscriber/lambda.cpp:29
ros2/examples/rclcpp/minimal_subscriber/not_composable.cpp:35
ros2/examples/rclcpp/minimal_subscriber/member_function.cpp:26
ros2/system_tests/test_cli_remapping/test/name_maker.cpp:31
ros2/system_tests/test_cli_remapping/test/name_maker.cpp:32
ros2/system_tests/test_cli_remapping/test/name_maker.cpp:33
ros2/demos/image_tools/src/cam2image.cpp:127
ros2/demos/image_tools/src/cam2image.cpp:146
ros2/demos/image_tools/src/showimage.cpp:152
ros2/demos/demo_nodes_cpp/src/topics/talker_serialized_message.cpp:118
ros2/demos/demo_nodes_cpp/src/topics/listener_serialized_message.cpp:79
ros2/demos/demo_nodes_cpp/src/topics/talker.cpp:61
ros2/demos/demo_nodes_cpp/src/topics/listener.cpp:53
ros2/demos/demo_nodes_cpp/src/topics/allocator_tutorial.cpp:173
ros2/demos/demo_nodes_cpp/src/topics/allocator_tutorial.cpp:178
ros2/demos/demo_nodes_cpp/src/topics/listener_best_effort.cpp:34
ros2/message_filters/test/test_fuzz.cpp:134
ros2/message_filters/include/message_filters/subscriber.h:169
ros2/message_filters/test/test_subscriber.cpp:68
ros2/message_filters/test/test_subscriber.cpp:87
ros2/message_filters/test/test_subscriber.cpp:106
ros2/message_filters/test/test_subscriber.cpp:129
ros2/message_filters/test/test_subscriber.cpp:152
ros2/message_filters/test/test_subscriber.cpp:176
ros2/message_filters/test/test_subscriber.cpp:216
ros2/message_filters/test/test_subscriber.cpp:234
ros2/message_filters/test/test_subscriber.cpp:254
ros2/message_filters/test/test_subscriber.cpp:256
ros2/message_filters/include/message_filters/subscriber.h:169
ros2/examples/rclcpp/minimal_composition/src/publisher_node.cpp:26
ros2/examples/rclcpp/minimal_composition/src/subscriber_node.cpp:26
ros2/demos/lifecycle/src/lifecycle_listener.cpp:42
ros2/demos/lifecycle/src/lifecycle_listener.cpp:49
ros2/demos/composition/src/talker_component.cpp:36
ros2/demos/composition/src/listener_component.cpp:46
ros2/demos/intra_process_demo/include/image_pipeline/watermark_node.hpp:39
ros2/demos/intra_process_demo/include/image_pipeline/watermark_node.hpp:60
ros2/demos/intra_process_demo/include/image_pipeline/camera_node.hpp:54
ros2/demos/intra_process_demo/include/image_pipeline/camera_node.hpp:54
ros2/demos/intra_process_demo/include/image_pipeline/image_view_node.hpp:65
ros2/demos/intra_process_demo/include/image_pipeline/watermark_node.hpp:39
ros2/demos/intra_process_demo/include/image_pipeline/watermark_node.hpp:60
ros2/demos/intra_process_demo/include/image_pipeline/image_view_node.hpp:65
ros2/demos/intra_process_demo/include/image_pipeline/camera_node.hpp:54
ros2/demos/intra_process_demo/include/image_pipeline/image_view_node.hpp:65
ros2/demos/intra_process_demo/include/image_pipeline/watermark_node.hpp:39
ros2/demos/intra_process_demo/include/image_pipeline/watermark_node.hpp:60
ros2/demos/intra_process_demo/src/two_node_pipeline/two_node_pipeline.cpp:33
ros2/demos/intra_process_demo/src/two_node_pipeline/two_node_pipeline.cpp:70
ros2/demos/intra_process_demo/src/cyclic_pipeline/cyclic_pipeline.cpp:33
ros2/demos/intra_process_demo/src/cyclic_pipeline/cyclic_pipeline.cpp:57
ros2/demos/pendulum_control/src/pendulum_demo.cpp:161
ros2/demos/pendulum_control/src/pendulum_demo.cpp:174
ros2/demos/pendulum_control/src/pendulum_demo.cpp:185
ros2/demos/pendulum_control/src/pendulum_demo.cpp:191
ros2/demos/pendulum_control/src/pendulum_demo.cpp:206
ros2/demos/pendulum_control/src/pendulum_demo.cpp:210
ros2/demos/pendulum_control/src/pendulum_teleop.cpp:49
ros2/demos/pendulum_control/src/pendulum_logger.cpp:81
ros2/geometry2/tf2_ros/src/static_transform_broadcaster.cpp:46
ros2/geometry2/tf2_ros/src/transform_broadcaster.cpp:45
ros2/geometry2/tf2_ros/src/transform_listener.cpp:84
ros2/geometry2/tf2_ros/src/transform_listener.cpp:86
ros2/geometry2/tf2_ros/src/tf2_monitor.cpp:167
ros2/geometry2/tf2_ros/src/tf2_monitor.cpp:169
ros2/geometry2/tf2_ros/test/message_filter_test.cpp:99
all/message_filters/include/message_filters/subscriber.h:169
ros2/robot_state_publisher/src/robot_state_publisher.cpp:92
ros2/robot_state_publisher/src/joint_state_listener.cpp:92
ros2/robot_state_publisher/src/joint_state_listener.cpp:92
ros2/system_tests/test_communication/test/test_publisher_subscriber_serialized.cpp:69
ros2/system_tests/test_communication/test/test_publisher_subscriber_serialized.cpp:71
ros2/system_tests/test_communication/test/test_publisher_subscriber_serialized.cpp:69
ros2/system_tests/test_communication/test/test_publisher_subscriber_serialized.cpp:71
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_publisher.cpp:35
ros2/system_tests/test_communication/test/test_subscription_valid_data.cpp:42
ros2/system_tests/test_communication/test/test_subscription_valid_data.cpp:47
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:99
ros2/system_tests/test_communication/test/test_publisher_subscriber.cpp:34
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_communication/test/test_subscriber.cpp:69
ros2/system_tests/test_rclcpp/test/test_timeout_subscriber.cpp:47
ros2/system_tests/test_rclcpp/test/test_intra_process.cpp:47
ros2/system_tests/test_rclcpp/test/test_intra_process.cpp:69
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:67
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:90
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:194
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:242
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:276
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:314
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:348
ros2/system_tests/test_rclcpp/test/./pub_sub_fixtures.hpp:49
ros2/system_tests/test_rclcpp/test/test_multithreaded.cpp:73
ros2/system_tests/test_rclcpp/test/test_multithreaded.cpp:292
ros2/system_tests/test_rclcpp/test/test_publisher.cpp:56
ros2/system_tests/test_rclcpp/test/./pub_sub_fixtures.hpp:49
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:50
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:52
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:77
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:79
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:67
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:90
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:194
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:242
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:276
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:314
ros2/system_tests/test_rclcpp/test/test_subscription.cpp:348
ros2/system_tests/test_rclcpp/test/./pub_sub_fixtures.hpp:49
ros2/system_tests/test_rclcpp/test/test_intra_process.cpp:47
ros2/system_tests/test_rclcpp/test/test_intra_process.cpp:69
ros2/system_tests/test_rclcpp/test/test_executor.cpp:211
ros2/system_tests/test_rclcpp/test/test_executor.cpp:216
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:50
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:52
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:77
ros2/system_tests/test_rclcpp/test/test_repeated_publisher_subscriber.cpp:79
ros2/system_tests/test_rclcpp/test/test_avoid_ros_namespace_conventions_qos.cpp:61
ros2/system_tests/test_rclcpp/test/./pub_sub_fixtures.hpp:49
ros2/system_tests/test_rclcpp/test/test_executor.cpp:211
ros2/system_tests/test_rclcpp/test/test_executor.cpp:216
ros2/system_tests/test_rclcpp/test/test_timeout_subscriber.cpp:47
ros2/system_tests/test_rclcpp/test/test_publisher.cpp:56
ros2/system_tests/test_rclcpp/test/./pub_sub_fixtures.hpp:49
ros2/system_tests/test_rclcpp/test/test_multithreaded.cpp:73
ros2/system_tests/test_rclcpp/test/test_multithreaded.cpp:292
ros2/system_tests/test_rclcpp/test/test_avoid_ros_namespace_conventions_qos.cpp:61
ros2/system_tests/test_rclcpp/test/./pub_sub_fixtures.hpp:49
ros2/rviz/rviz_default_plugins/src/rviz_default_plugins/displays/camera/camera_display.cpp:291
ros2/rviz/rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp:196
ros2/rviz/rviz_default_plugins/src/rviz_default_plugins/displays/marker/marker_display.cpp:84
ros2/rviz/rviz_default_plugins/src/rviz_default_plugins/tools/nav_goal/goal_tool.cpp:69
ros2/rviz/rviz_default_plugins/src/rviz_default_plugins/tools/pose_estimate/initial_pose_tool.cpp:68
ros2/rviz/rviz_default_plugins/src/rviz_default_plugins/tools/point/point_tool.cpp:83
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/pose/../../publishers/pose_publisher.hpp:50
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/range/../../publishers/range_publisher.hpp:53
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/relative_humidity/../../publishers/relative_humidity_publisher.hpp:65
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/pose_array/../../publishers/pose_array_publisher.hpp:50
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/pointcloud/../../publishers/point_cloud2_publisher.hpp:68
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/temperature/../../publishers/temperature_publisher.hpp:65
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/pointcloud/../../publishers/point_cloud_publisher.hpp:67
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/marker/../../publishers/marker_publisher.hpp:59
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/marker/../../publishers/marker_publisher.hpp:68
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/marker/../../publishers/marker_array_publisher.hpp:60
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/map/../../publishers/map_publisher.hpp:54
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/map/../../publishers/single_marker_publisher.hpp:57
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/odometry/../../publishers/odometry_publisher.hpp:57
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/laser_scan/../../publishers/laser_scan_publisher.hpp:55
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/illuminance/../../publishers/illuminance_publisher.hpp:65
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/marker_array/../../publishers/marker_publisher.hpp:59
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/marker_array/../../publishers/marker_publisher.hpp:68
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/marker_array/../../publishers/marker_array_publisher.hpp:60
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/point/../../publishers/point_stamped_publisher.hpp:50
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/camera/../../publishers/camera_info_publisher.hpp:49
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/camera/../../publishers/image_publisher.hpp:63
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/camera/../../publishers/point_cloud_publisher.hpp:67
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/path/../../publishers/path_publisher.hpp:54
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/fluid_pressure/../../publishers/fluid_pressure_publisher.hpp:65
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/grid_cells/../../publishers/grid_cells_publisher.hpp:52
ros2/rviz/rviz_default_plugins/test/rviz_default_plugins/displays/image/../../publishers/image_publisher.hpp:63
```

emersonknapp commented:
Another note - `[[deprecated]]` causes the cppcheck linter check to crash. The printed stack trace/error is unhelpful (see pasted below) - but some digging shows that it is reporting the `[[deprecated]]` as invalid syntax. I did some investigation into this and it seems that we are using a version of cppcheck that supposedly added support for this attribute, so I'm not sure why it can't handle it.

```                                                                                                                                                                        
34: -- run_test.py: invoking following command in '/root/ros/src/ros2/rclcpp/rclcpp':                                                                                                                 
34:  - /root/ros/install/ament_cppcheck/bin/ament_cppcheck --xunit-file /root/ros/build/rclcpp/test_results/rclcpp/cppcheck.xunit.xml --include_dirs /root/ros/src/ros2/rclcpp/rclcpp/include
34: Traceback (most recent call last):                                                                      
34:   File "/root/ros/install/ament_cppcheck/bin/ament_cppcheck", line 11, in <module>                                                                                                                                               
34:     load_entry_point('ament-cppcheck==0.6.3', 'console_scripts', 'ament_cppcheck')()                                                                                                                                             
34:   File "/root/ros/install/ament_cppcheck/lib/python3.6/site-packages/ament_cppcheck/main.py", line 128, in main                                                                       
34:     report[filename].append(data)                                                                  
34: KeyError: '/root/ros/src/ros2/rclcpp/rclcpp/include/rclcpp/node.hpp'   
```

emersonknapp commented:
@wjwwood do you have thoughts on how we should move forward based on this data? CI will just report failed for rviz_default_plugins

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -284,7 +285,7 @@ class Node : public std::enable_shared_from_this<Node>
     CallbackT && callback,
     size_t qos_history_depth,
-    rclcpp::callback_group::CallbackGroup::SharedPtr group = nullptr,
+    rclcpp::callback_group::CallbackGroup::SharedPtr group,
```
This breaks API though right? You cannot do this to resolve the ambiguity.

emersonknapp replied:
I'm not sure I follow what you mean by "break the api" here. I have redirected existing calls to `topic_name, callback, qos_history_depth` through the new API - so they're compiling against a new function. But in `node_impl.hpp` I have directed all three interfaces through the new Options-based interface, so they do all receive the same functionality.

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -171,8 +171,9 @@ class Node : public std::enable_shared_from_this<Node>
-    std::shared_ptr<Alloc> allocator = nullptr,
+    const std::string & topic_name,
+    size_t qos_history_depth,
+    std::shared_ptr<Alloc> allocator,
```
Same here. This breaks API.

wjwwood commented:
That's a lot of instances, which implies the existing user code will be similar, so I'd say leave them as deprecated and we'll remove them in the next release. Then fix all the instances you found to use the new signature as part of merging this pull request.

With respect to the cppcheck issue, that's odd because we're using that elsewhere already. Maybe try giving a message with the deprecation, e.g. `[[deprecated("use X instead")]]` which is also better practice (if you deprecate something you should suggest what to do instead).

wjwwood commented:
For the API break, I'd recommend making the signature which takes the options structure not have a default argument for the options, which changes the number of required arguments to three. In the future, when we remove the deprecated methods we can re-add the default argument for the option struct, making it the new call when you use just topic name and history depth.

emersonknapp commented:
I don't think I understand/agree with that rollout plan - the simpler version of the API is marked deprecate, that means that proactive users will start explicitly passing in SubscriptionOptions/PublisherOptions to make the deprecation warnings go away - but on the next release when we add the default back for those arguments, their existing code would have been fine again

wjwwood commented:
Sorry, I think you're right, now that I think about it more. I think I was concerned that users will now be implicitly calling a different method than they were before without any changes in their code, however, the behavior of the new method they're calling should have the same behavior, so long as they're just passing the require arguments in the first place.

Sorry for the noise, I think your proposed diff is ok.

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -160,10 +168,12 @@ class Node : public std::enable_shared_from_this<Node>
   template<
     typename MessageT, typename Alloc = std::allocator<void>,
     typename PublisherT = ::rclcpp::Publisher<MessageT, Alloc>>
+  [[deprecated]]
```
The deprecation warning should read something like:

> "use the `create_publisher(const std::string &, size_t, const PublisherOptions<Alloc> & = PublisherOptions<Alloc>())` signature instead"

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -176,13 +186,43 @@ class Node : public std::enable_shared_from_this<Node>
   template<
     typename MessageT, typename Alloc = std::allocator<void>,
     typename PublisherT = ::rclcpp::Publisher<MessageT, Alloc>>
+  [[deprecated]]
```
The deprecation warning should read something like:

> "use the `create_publisher(const std::string &, size_t, const PublisherOptions<Alloc> & = PublisherOptions<Alloc>())` signature instead"

This deprecation I think is still good to do, but I would also be ok to drop this one for this pr, and do it in a follow up pr. Because at the same time we should also provide a way to let users specify even more QoS settings. If we just deprecate this, then I don't think you can, for instance, turn on best effort without invoking a deprecate function.

So let's just drop this deprecation from this pr, despite my previous request, and that should also reduce considerably the number of packages you need to touch to merge this.

emersonknapp replied:
Rephrasing for my own clarity - We will not deprecate the API that does not have a mandatory `qos_history_depth`, so that the following usage type will be ok in this pass.

```
create_subscription(topic, callback)
create_publisher(topic)
```

Meaning remove the `[[deprecated]]` from this func and the matching `create_subscription`

wjwwood replied:
Correct!

I'll deprecate it in a different pr, when I also add an option to pass an rclcpp specific QoS struct, which can only be created by providing a history depth, and then I'd feel ok deprecating the signature that takes the rmw QoS struct.

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -204,6 +244,7 @@ class Node : public std::enable_shared_from_this<Node>
     typename Alloc = std::allocator<void>,
     typename SubscriptionT = rclcpp::Subscription<
       typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>>
+  [[deprecated]]
```
Same rationale as the publisher equivalent, drop this deprecation for now.

emersonknapp commented:
So do you still propose that we update the 23 packages across ~9 repos in order to add this new API?

I am wondering if, before investing that effort, we should instead phase into that change -
1. Add the new API, don't deprecate anything
2. Deprecate the old APIs and change the places in the core that are affected

That would let us stabilize and review the new QoS features against this Options API, without blocking that on fixing up a bunch of other unrelated packages.

We could still put in the push to deprecate before the API freeze in a few weeks, but unblock the other threads of work in the meantime.

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -238,12 +279,13 @@ class Node : public std::enable_shared_from_this<Node>
     typename Alloc = std::allocator<void>,
     typename SubscriptionT = rclcpp::Subscription<
       typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>>
+  [[deprecated]]
```
The deprecation warning should read something like:

> "use the `create_subscription(const std::string &, CallbackT &&, size_t, const SubscriptionOptions<Alloc> & = SubscriptionOptions<Alloc>(), ...)` signature instead"

wjwwood commented on rclcpp/include/rclcpp/subscription_options.hpp:
```diff
@@ -0,0 +1,41 @@
+  bool ignore_local_publications = false;
+  rclcpp::callback_group::CallbackGroup::SharedPtr callback_group = nullptr;
+  std::shared_ptr<Alloc> allocator = nullptr;
+  IntraProcessSetting use_intra_process_comm = IntraProcessSetting::NodeDefault;
```
Some documentation on what each of these does would be nice, since they're no longer documented along with the method that takes it, same for the publisher options. See the NodeOptions as an example. Though it doesn't necessarily need getters and setters like it has, because it doesn't have resources to protect, they would be nice, so that they parameter idiom could be used with these option structures. However, that can be deferred a future improvement.

emersonknapp replied:
Will add the documentation.

I prefer to defer getters/setters to a future improvement to keep this change focused.

wjwwood replied:
> I prefer to defer getters/setters to a future improvement to keep this change focused.

Agreed.

wjwwood commented on rclcpp/test/test_pub_sub_option_interface.cpp:
```diff
@@ -0,0 +1,70 @@
+
+  auto topic_only_sub = node->create_subscription<test_msgs::msg::Empty>(
+    "topic_only",
+    [](std::shared_ptr<test_msgs::msg::Empty> test_msg) {(void) test_msg;});
```
Note for these, you can write the lambda as `[](std::shared_ptr<test_msgs::msg::Empty>) {}`.

emersonknapp replied:
Nice, thanks

wjwwood commented:
> That would let us stabilize and review the new QoS features against this Options API, without blocking that on fixing up a bunch of other unrelated packages.

I would rather see the new functions and the deprecations at the same time. But the number of places you need to touch should go down considerably if you remove the deprecation I suggested in my most recent comments.

wjwwood commented on rclcpp/include/rclcpp/subscription_options.hpp:
```diff
@@ -0,0 +1,46 @@
+{
+
+/// Structure containing optional configuration for Subscriptions.
+template<typename Alloc = std::allocator<void>>
```
I notice this, when looking at the changes in other packages, can you use `Allocator` here please? I know we use `Alloc` in a few places, but I'd like to change those too at some point, using the full name avoids abbreviation and brings it inline with the style the standard library uses, e.g. `std::base_string`: `class Allocator = std::allocator<CharT>` (https://en.cppreference.com/w/cpp/string/basic_string)

emersonknapp replied:
Ok - I have updated to use the non-abbreviated version in everything I added in this PR

emersonknapp commented:
Updated - this latest only threw deprecated in 3 packages - which now have open PRs "connected to" this one. Going to run CI against the bundle of PRs

@thomas-moulard - please run the following CI job:
* Gist file: https://gist.githubusercontent.com/emersonknapp/24c71173066cc13ae023be0c310807a8/raw/504e3fc2a89b8bc7349bfc5d156abbb5fcbef0d9/options-api.repos
* Additional BUILD flags: (note - no package selector, explicitly trying to build everything)
* Additional TEST flags: --packages-select rclcpp test_rclcpp tlsf_cpp demo_nodes_cpp
*  CI Job: ci_launcher

wjwwood commented:
I got the CI for you:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6665)](http://ci.ros2.org/job/ci_linux/6665/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2985)](http://ci.ros2.org/job/ci_linux-aarch64/2985/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5472)](http://ci.ros2.org/job/ci_osx/5472/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6494)](http://ci.ros2.org/job/ci_windows/6494/)

emersonknapp commented:
@wjwwood I'm not positive but it looks like the whole build farm borked out during that run?

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6673)](http://ci.ros2.org/job/ci_linux/6673/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2992)](http://ci.ros2.org/job/ci_linux-aarch64/2992/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5481)](http://ci.ros2.org/job/ci_osx/5481/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6502)](http://ci.ros2.org/job/ci_windows/6502/)

emersonknapp commented:
@wjwwood So far the build looks good - except for the `cppcheck` issue I mentioned. I did a little isolation test - it looks like it doesn't like `[[deprecated]]` on templated functions.

```
/**
  * test.cpp
  */
1  #include <cassert>
2 
3  [[deprecated]]
4  int ok() {
5    return 1;
6  }
7 
8  template<typename T>
9  [[deprecated]]
10 T not_ok(T a, T b) {
11   return a + b;
12 }
13 
14 
15 int main() {
16   int a = ok();
17   int b = not_ok(a, a);
18   assert(a == 1);
19   assert(b == 2);
20   return 0;
21 }


/**
  * build
  */
root:~/ros# g++ test.cpp
test.cpp: In function ‘int main()’:
test.cpp:16:14: warning: ‘int ok()’ is deprecated [-Wdeprecated-declarations]
   int a = ok();
              ^
test.cpp:4:5: note: declared here
 int ok() {
     ^~
test.cpp:17:22: warning: ‘T not_ok(T, T) [with T = int]’ is deprecated [-Wdeprecated-declarations]
   int b = not_ok(a, a);
                      ^
test.cpp:10:3: note: declared here
 T not_ok(T a, T b) {
   ^~~~~~


/**
  * lint
  */
root@d0f8c7728dfc:~/ros# /root/ros/install/ament_cppcheck/bin/ament_cppcheck test.cpp
[test.cpp:9]: (error: syntaxError) syntax error
1 errors
```

wjwwood commented:
Does the `// cppcheck-suppress ...` feature able to workaround this syntax error? (see: http://cppcheck.sourceforge.net/manual.pdf)

If not we'll have to disable cppcheck on rclcpp until we can figure out a solution.

Is there an issue on cppcheck's issue tracker? If not this minimal example may be useful to them.

emersonknapp commented:
Ah - sorry - I updated my comment above with more info - this was fixed in cppcheck 1.83

emersonknapp commented:
The suppress syntax works, thanks for pointing me to that. I have updated to suppress the error.

Oddly, it was only required on one of the two signatures that used `[[deprecated]]`, I tried fiddling around but wasn't able to tell what the difference was between them, I'm guessing it's some subtle bug in cppcheck's parser.

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6677)](http://ci.ros2.org/job/ci_linux/6677/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2996)](http://ci.ros2.org/job/ci_linux-aarch64/2996/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5485)](http://ci.ros2.org/job/ci_osx/5485/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6507)](http://ci.ros2.org/job/ci_windows/6507/)

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6677)](http://ci.ros2.org/job/ci_linux/6677/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2996)](http://ci.ros2.org/job/ci_linux-aarch64/2996/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5485)](http://ci.ros2.org/job/ci_osx/5485/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6507)](http://ci.ros2.org/job/ci_windows/6507/)
